core/iwasm/libraries/wasi-nn/src/utils/wasi_nn_app_native.c

```diff
@@ -72,6 +72,51 @@ graph_builder_array_app_native(wasm_module_inst_t instance,
     return success;
 }
 
+#if WASM_ENABLE_WASI_EPHEMERAL_NN != 0
+error
+graph_builder_array_app_native_ephemeral(wasm_module_inst_t instance,
+                                         graph_builder_wasm *builder_wasm,
+                                         uint32_t size,
+                                         graph_builder_array *builder_array)
+{
+    if (!wasm_runtime_validate_native_addr(instance, builder_wasm,
+                                           size * sizeof(graph_builder_wasm))) {
+        NN_ERR_PRINTF("builder_wasm is invalid");
+        return invalid_argument;
+    }
+
+    NN_DBG_PRINTF("Graph builder array contains %d elements", size);
+
+    graph_builder *builder =
+        (graph_builder *)wasm_runtime_malloc(size * sizeof(graph_builder));
+    if (builder == NULL)
+        return missing_memory;
+
+    for (uint32_t i = 0; i < size; ++i) {
+        if (!wasm_runtime_validate_app_addr(
+                instance, builder_wasm[i].buf_offset, builder_wasm[i].size)) {
+            NN_ERR_PRINTF("builder_wasm[i].buf_offset is invalid");
+            return invalid_argument;
```

Comment:
Should `wasm_runtime_free(builder)` before return

Reply:
Sorry, I forgot it.